feat(logplex source): Introduce Heroku button

Hoverbear opened the pull request:
Fixes #2887 

Hard blocked by: #2922 
Related: #2921 #2919 

Adds Heroku Deploy button for Vector. 

Some notes:

* It takes a URL and auto refreshes from that URL each boot. Users can opt out by provisioning a `./vector.toml`.
* It pulls the newest stable musl build of vector, if it doesn't exist. This means users always get the new stable vector, and it doesn't update unless the node is take down and re-provisioned.
* It relies on HTTPS Basic Auth to secure the source, (Not implemented in the buildpack yet!) but it's blocked by #2922 .

Further work:

* FIx #2922
* Add HTTPS basic auth to the vector template in the buildpack. (See https://devcenter.heroku.com/articles/log-drains#https-drains for auth)
* The address needs to be changed to point to master not this branch
* @binarylogic level docs pass

Hoverbear commented:
@binarylogic Let me know if you want me to get this ready for a review/merge again. :)

binarylogic commented on app.json:
```diff
@@ -0,0 +1,50 @@
+            "description": "The URL of the `vector.toml` you use to use. (Refreshed on restart! For static, provision a `vector.toml` with your app.)",
+            "required": false
+        },
+        "AWS_ACCESS_KEY_ID": {
+            "description": "An AWS key ID (if needed, for sinking to CloudWatch)",
+            "required": false
+        },
+        "AWS_SECRET_ACCESS_KEY": {
+            "description": "An AWS access key (if needed, for sinking to CloudWatch)",
+            "required": false
+        },
+        "AWS_REGION": {
+            "description": "An AWS region (if needed, for sinking to CloudWatch). Examples: `ca-central-1`, `us-east-1`",
+            "required": false
+        },
+        "AWS_CLOUDWATCH_STREAM_NAME": {
+            "description": "An AWS CloudWatch stream name (if needed, for sinking to CloudWatch. Vector can create this for you.)",
+            "required": false
+        },
+        "AWS_CLOUDWATCH_GROUP_NAME": {
+            "description": "An AWS CloudWatch stream group name (if needed, for sinking to CloudWatch. You need to create this yourself.)",
+            "required": false
+        },
```
@Hoverbear are these necessary given that we no longer prefer Cloudwatch? This is really the big thing for me. I want to make sure deploying Vector as a Heroku button is not restricted to AWS or Cloudwatch.

Hoverbear replied:
We can totally yank them! :) Where would you like the logs to go then?

binarylogic replied:
Couldn't we base it entirely on the remote Vector config?

Hoverbear replied:
We could! Yeah! Okay! I'll refactor this :)

jamtur01 commented:
@binarylogic How do you want to proceed here?

binarylogic commented:
If the remaining work is <= 3 then let's finish it. I just want to make this generic and rip out any AWS/CloudWatch assumptions.

Hoverbear commented:
https://github.com/timberio/vector/issues/2922 is required, I think it's around a 3 unless there is something I'm not seeing. I removed cloudwatch assumptions.

jszwedko commented:
If we are just letting people specify arbitrary vector configs, it doesn't seem like this would be blocked by https://github.com/timberio/vector/issues/2922 . Specifically using a Heroku deployed vector to consume logplex logs would be, but not generally deploying vector onto Heroku.

Hoverbear commented:
Indeed, this is more of a soft-block, but I would be very uncomfortable with someone accidentally leaving their logging router open to the public because we lacked a feature. :(

jszwedko commented:
True, I'm thinking that they could deploy it using sources that behave like consumers, like the `kafka` source, though.

We should definitely add basic auth support too.

Hoverbear commented:
Some notes:

* Since we no longer have a default to cloudwatch I moved this to a blackhole.
* Since we have no reasonable default I felt it was required to write more docs.
* Added basic auth support.
* Kept remote URL support but no longer make it recommended. It's preferred if users push to the repo.

This is good for a final review. @jamtur01

Hoverbear commented:
Oh! Last note, this should wait till 0.11 is out.

jamtur01 commented:
@Hoverbear Now 0.11 is out - can this proceed?

Hoverbear commented:
Heck yeah!
